src/connections/allContacts.js

```diff
@@ -1,5 +1,49 @@
 import { connect, withMutations } from 'cozy-client'
 import flow from 'lodash/flow'
+import array from 'lodash/array'
+import isEqual from 'lodash/isEqual'
+
+function findContacts(keys, attributes) {
+  return client =>
+    Promise.all(
+      keys.map(async ({ attr, prop }) => {
+        const result = await getExistingContactsByProp(
+          attributes[attr],
+          attr,
+          prop,
+          client
+        )
+        return result.data
+      })
+    )
+      .then(results => [].concat(...results))
+      .then(results => array.uniqWith(results, isEqual))
+}
+
+function getExistingContactsByProp(props = [], attr, prop, client) {
```

Comment:
I think the responsibilities of this function are a bit too large.

I think decoupling between the creation of the selectors and the execution of the query would lead to easier testing.

Moreover, I think the creation of the selectors could be something very interesting to include inside a helper library like `cozy-client` or `cozy-doctypes` (recently moved in cozy-libs).

It could be something like this : 

```js
/* Creates a selector from a document and a path inside this document */
const createMatcher = (doc, path) => {
    const attrs = path.split('.')
    let cur = doc
    let selector = {}
    for (let i = 0; i < attrs.length; i++) {
        const attr = attrs[i]
        cur = cur[attr]
        if (Array.isArray(cur)) {
            if (i == attrs.length - 1) {
                throw new Error('Cannot have last attr an an array')
            }
            const nextAttr = attrs[i + 1]
            const values = cur.map(x => x[nextAttr])
            sel[attr] = {
                $elemMatch: {
                    [nextAttr]: {
                        $in: values
                    }
                }
            }
        } else {
            sel[attr] = cur
        }
    }
    return selector
}

const allContacts = new Query.all('io.cozy.contacts')

const findSimilarContacts = (contact, client) => {
    const selectors = [
        createMatcher(contact, 'phone.number'),
        createMatcher(contact, 'email.address')
    ]
    const results = flatMap(await Promise.all(
        selectors.map(
            selector => client.query(allContacts.where(selector)
        )
    ))
    return uniqWith(results, isEqual)
} 
```

What do you think ?

Reply:
You are absolutly right! More work needs to be made to generalize and explicit this part of code.

Reply:
Find a new attempt here: https://github.com/cozy/cozy-contacts/pull/241/commits/ab4ca90c0bf0d2f012b046c36acc433cabe7624a